src/signed_duration.rs

```diff
@@ -0,0 +1,2189 @@
+        // never overflows i32.
+        let nanos = (millis % MILLIS_PER_SEC) as i32 * NANOS_PER_MILLI;
+        SignedDuration::new_unchecked(secs, nanos)
+    }
+
+    /// Creates a new `SignedDuration` from the given number of whole
+    /// microseconds.
+    ///
+    /// Note that since this accepts an `i64`, this method cannot be used
+    /// to construct the full range of possible signed duration values. In
+    /// particular, [`SignedDuration::as_micros`] returns an `i128`, and this
+    /// may be a value that would otherwise overflow an `i64`.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use jiff::SignedDuration;
+    ///
+    /// let duration = SignedDuration::from_micros(12_000_456);
+    /// assert_eq!(duration.as_secs(), 12);
+    /// assert_eq!(duration.subsec_nanos(), 456_000);
+    ///
+    /// let duration = SignedDuration::from_micros(-12_000_456);
+    /// assert_eq!(duration.as_secs(), -12);
+    /// assert_eq!(duration.subsec_nanos(), -456_000);
+    /// ```
+    #[inline]
+    pub const fn from_micros(micros: i64) -> SignedDuration {
+        // OK because MICROS_PER_SEC!={-1,0}.
+        let secs = micros / MICROS_PER_SEC;
+        // OK because MICROS_PER_SEC!={-1,0} and because
+        // millis % MICROS_PER_SEC can be at most 999, and 999 * 1_000_000
+        // never overflows i32.
+        let nanos = (micros % MICROS_PER_SEC) as i32 * NANOS_PER_MICRO;
+        SignedDuration::new_unchecked(secs, nanos)
+    }
+
+    /// Creates a new `SignedDuration` from the given number of whole
+    /// nanoseconds.
+    ///
+    /// Note that since this accepts an `i64`, this method cannot be used
+    /// to construct the full range of possible signed duration values. In
+    /// particular, [`SignedDuration::as_nanos`] returns an `i128`, and this
+    /// may be a value that would otherwise overflow an `i64`.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use jiff::SignedDuration;
+    ///
+    /// let duration = SignedDuration::from_nanos(12_000_000_456);
+    /// assert_eq!(duration.as_secs(), 12);
+    /// assert_eq!(duration.subsec_nanos(), 456);
+    ///
+    /// let duration = SignedDuration::from_nanos(-12_000_000_456);
+    /// assert_eq!(duration.as_secs(), -12);
+    /// assert_eq!(duration.subsec_nanos(), -456);
+    /// ```
+    #[inline]
+    pub const fn from_nanos(nanos: i64) -> SignedDuration {
+        // OK because NANOS_PER_SEC!={-1,0}.
+        let secs = nanos / (NANOS_PER_SEC as i64);
+        // OK because NANOS_PER_SEC!={-1,0}.
+        let nanos = (nanos % (NANOS_PER_SEC as i64)) as i32;
+        SignedDuration::new_unchecked(secs, nanos)
+    }
+
+    /// Creates a new `SignedDuration` from the given number of hours. Every
+    /// hour is exactly `3,600` seconds.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the number of hours, after being converted to nanoseconds,
+    /// overflows the minimum or maximum `SignedDuration` values.
```

Comment:
Shall we then provide a fallible constructor like `try_from_hours`? So does `from_mins`.

Reply:
There aren't fallible constructors for any units. That matches std. The design philosophy is for SignedDuration to match std as closely as possible. I'm open to adding fallible constructors, but I want to see compelling use cases.

Reply:
Thanks for your explanation. I noticed that the upstream APIs are unstable also, so let's see what's the consensus - https://github.com/rust-lang/rust/issues/120301/

Both fallible constructor and ambiguous semantic are thrown. I noticed an argument that `hours` and `minutes` are not supported for leap seconds and more ([ref](https://rust-lang.github.io/rfcs/1040-duration-reform.html)), but in jiff we typically consider days and above are ambiguous? Is there certain different consideration?

Reply:
Sorry, I don't understand your question. Jiff specifically and intentionally doesn't support leap seconds. And whether leap seconds, even if supported, should impact the definition of hours or minutes is unclear.

Reply:
> doesn't support leap seconds

OK. Then it looks some concerns from the [Rust RFC 1040](https://rust-lang.github.io/rfcs/1040-duration-reform.html) proposer.

> even if supported, should impact the definition of hours or minutes is unclear

Yeah .. So the reason jiff don't support days and above to `SignedDuration`, is that:

* How many days in a year is indeterminate (leap year)
* How many days in a month is indeterminate (of course)
* How many hours in a day is indeterminate (I'm not sure but perhaps daylight savings things?)
* How many hours in a week is indeterminate (because of day -> hours is indeterminate)

Is this understanding properly?

Reply:
Yes

Reply:
I would definitely encourage you to read more of Jiff's docs. And especially the examples. They'll demonstrate a lot of these weird edge cases.

Reply:
> I would definitely encourage you to read more of Jiff's docs. And especially the examples. They'll demonstrate a lot of these weird edge cases.

Sure .. Thanks for providing such detailed docs.